### ISSUE TITLE
How to freeze the corpus model and fine-tune query model?

### ISSUE THREAD

**Issue by ykhatami** [UKPLab/sentence-transformers#1543]
Hello, in the bi-encoder setup for question answering, I want to start from one model e.g. all-MiniLM-L12-v2, for both query and corpus models. Then I want to only fine-tune the query model while the corpus model is frozen. 

One way is to go through torch, and implement the models there. I'm wondering if this is readily supported in sentence-transformers. Is there a module that supports this? Is there any script you could share? 

Thanks

**Comment by ykhatami** [comment]
I saw there is an asymmetric option at https://github.com/UKPLab/sentence-transformers/releases/tag/v0.4.1
The starting point is from a base model models.Transformer(base_model, max_seq_length=250). Can I use a pre-trained model from sentence-transformers like all-MiniLM-L12-v2 here? How would I treat the pooling layer applied at asymmetric models example?

**Comment by ykhatami** [comment]
I found a solution that works for training. Would be great if you can confirm this is the right thing to do:
```
InputExample(texts=[{'QRY':'This is newspaper'}, {'DOC':'that was in the news' }], label=1.0)
```
Instead of passing 1 dict, pass 2 dicts. One thing to note is that this example cannot be passed to model.encode(). For that we need to pass 1 dict with 2 keys 'QRY' and 'DOC'.

**Comment by nreimers** [comment]
Yes, the approach looks right

**Comment by ykhatami** [comment]
Thanks for confirming. You may want to update this page
https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/models/Asym.py

`train_example = InputExample(texts=[{'query': 'Train query', 'doc': 'Doc query'}], label=1)`
to
`train_example = InputExample(texts=[{'query': 'Train query'}, {'doc': 'Doc query'}], label=1)`

**Comment by ykhatami** [comment]
Resolved

**Comment by nreimers** [comment]
@ykhatami Thanks for pointing this out. Updated that file

**Comment by emilysilcock** [comment]
@ykhatami this is very similar to a problem that I am working with. Thanks for sharing the code above! 
I was wondering whether this does actually lead to the query model being frozen, or whether both models are trained, just asymmetrically?
